### PR TITLE
39 installer for sharkcage

### PR DESCRIPTION
**Pull request by SailReal**
* Install executable to an specific dir
* Set path in registry
* Create shortcuts on desktop and in taskbar
* Install and start service

### CODE REVIEW

File: SharkCage/CageService/CageService.vcxproj [hunk 1]
```diff
@@ -178,6 +178,7 @@ xcopy /y "$(OutDir)$(TargetName).pdb" "$(SolutionDir)post_build_output\"
       <SDLCheck>true</SDLCheck>
       <DisableSpecificWarnings>4834</DisableSpecificWarnings>
       <TreatWarningAsError>true</TreatWarningAsError>
+      <AdditionalOptions>/std:c++latest %(AdditionalOptions)</AdditionalOptions>
```
Comment (DonatJR) [inline SharkCage/CageService/CageService.vcxproj]: Just curious: why is this needed if nothing else in the CageService project changed?

File: SharkCage/SharkCage.sln [hunk 2]
```diff
@@ -33,8 +37,8 @@ Global
 	GlobalSection(ProjectConfigurationPlatforms) = postSolution
 		{569449B5-18D8-43F0-A6A0-7E754EFEBC25}.Debug|Any CPU.ActiveCfg = Debug|Win32
 		{569449B5-18D8-43F0-A6A0-7E754EFEBC25}.Debug|Any CPU.Build.0 = Debug|Win32
-		{569449B5-18D8-43F0-A6A0-7E754EFEBC25}.Debug|x64.ActiveCfg = Debug|Win32
-		{569449B5-18D8-43F0-A6A0-7E754EFEBC25}.Debug|x64.Build.0 = Debug|Win32
+		{569449B5-18D8-43F0-A6A0-7E754EFEBC25}.Debug|x64.ActiveCfg = Release|Win32
+		{569449B5-18D8-43F0-A6A0-7E754EFEBC25}.Debug|x64.Build.0 = Release|Win32
```
Comment (DonatJR) [inline SharkCage/SharkCage.sln]: Debug + x64 configuration now builds the Release version of the service again...
image

File: SharkCage/CageServiceInstaller/CageServiceInstaller.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        public override void Rollback(IDictionary saved_state)
+        {
+            base.Rollback(saved_state);
+            UninstallService();
+        }
+
+        public override void Uninstall(System.Collections.IDictionary saved_state)
+        {
+            base.Uninstall(saved_state);
+            UninstallService();
```
Comment (DonatJR) [inline SharkCage/CageServiceInstaller/CageServiceInstaller.cs]: Could it be changing the order of uninstalling is as easy as switching these two lines? (So the installer won't show the running service message)

File: SharkCage/CageServiceInstaller/ServiceInstaller.cs [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+            }
+            finally
+            {
+                CloseServiceHandle(scm);
+            }
+        }
+
+        private static void StartService(IntPtr service)
+        {
+            SERVICE_STATUS status = new SERVICE_STATUS();
```
Comment (DonatJR) [inline SharkCage/CageServiceInstaller/ServiceInstaller.cs]: what is the status used for?

File: SharkCage/CageServiceInstaller/ServiceInstaller.cs [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+        private static void StopService(IntPtr service)
+        {
+            SERVICE_STATUS status = new SERVICE_STATUS();
+            ControlService(service, ServiceControl.Stop, status);
+
+            var amount_of_retries = 5;
+            var changed_status = false;
+
+            // It takes sometimes longer to stop the service than the dwWaitHint is set by the system.
+            for (; amount_of_retries > 0; amount_of_retries--)
```
Comment (DonatJR) [inline SharkCage/CageServiceInstaller/ServiceInstaller.cs]: is there a reason `amount_of_retries` is not initialized in the for loop directly?

File: SharkCage/CageService/CageService.vcxproj (hunk 1 above)
Comment (SailReal) [reply SharkCage/CageService/CageService.vcxproj]: Hmm I think this line was added by mistake...tried to get the release build working, but isn't needed anymore.

File: SharkCage/CageServiceInstaller/CageServiceInstaller.cs (hunk 3 above)
Comment (SailReal) [reply SharkCage/CageServiceInstaller/CageServiceInstaller.cs]: Haha had the same idea but unfortunately not, the complete functions is called after popping up the running service message (and pressing continue...).

**Comment by SailReal** [comment]
> I can't open (/build / test) the installer project because my VS is missing some needed part for it, can you tell me what the prerequisites are for the installer project?

Sorry for that, you have to install an additional project type "Visual Studio Installer".
screen shot 2018-06-30 at 11 04 51 am

File: SharkCage/CageServiceInstaller/ServiceInstaller.cs (hunk 4 above)
Comment (SailReal) [reply SharkCage/CageServiceInstaller/ServiceInstaller.cs]: The status is required for the `StartService` dll function. But checking the status right after finishing the function doesn't make sense because of the system needs some time to change the status of the service. Because of that this is handled in the `WaitForServiceStatus` function.

But I optimized the call a bit: `ControlService(service, ServiceControl.Stop, new SERVICE_STATUS());`

**Comment by SailReal** [comment]
Hmm stay tuned 🍿, too long ago, will figure out which component is required!

File: SharkCage/CageService/CageService.vcxproj (hunk 1 above)
Comment (SailReal) [reply SharkCage/CageService/CageService.vcxproj]: I looked at it again, without the `/std:c++latest %(AdditionalOptions)` the `optional` is not available in the CageService, but used:
`fatal error C1189: #error:  class template optional is only available with C++17.`

**Comment by SailReal** [comment]
I'm, quiet sure that I installed this one: [MicrosoftVisualStudio2017InstallerProjects](https://marketplace.visualstudio.com/items?itemName=VisualStudioProductTeam.MicrosoftVisualStudio2017InstallerProjects)

**Comment by DonatJR** [comment]
Just thought about the /silent option to install the service without user interaction. Do you think this should be part of this PR?

**Comment by SailReal** [comment]
Good points, both (x86/x64 and silent installer) should be part of this PR 👍

**Comment by DonatJR** [comment]
Quiet install / install without user input is already baked into all msi installers, we just have to use a tool called `msiexec` shipped with windows. For example: `msiexec /i SharkCage.msi /qb+` just displays a progress bar and informs the user on success. 
image

Building the installer for AnyCPU and x64 should be as easy as ticking the build-box displayed above. Once this is done I will approve the PR. :)

**Comment by SailReal** [comment]
Should be fixed now, thx!

**Comment by SailReal** [comment]
Thanks for the review!!!
